### PR TITLE
Modernize

### LINKED ISSUE #222: Add "I don't support Internet Explorer" to the list of versions
I really enjoy using this tool to see what native APIs I can use over jQuery. However, on my side projects, I don't really ever think about supporting Internet Explorer. Could there be an option to showcase newer, more experimental features of browsers (async/await, modules, fetch API, etc.) that are much newer/more up-to-date than anything IE has?

### LINKED ISSUE #247: Range labels should be clickable for accessibility

### LINKED ISSUE #266: Modernization
Taken from https://github.com/HubSpot/youmightnotneedjquery/pull/249#issuecomment-1229298952, but I didn't want it to get lost from other contributors.

- [x] Convert CoffeeScript to JavaScript
- [x] Convert Stylus to CSS with nesting working draft (would mostly just be adding brackets like JS)
- [x] Convert Jade to something more HTML based (like EJS or React/JSX)
- [x] Remove Gulp in favor of a modern bundler (like esbuild, Vite, or swc), native web code, or a static site generator (like Astro)
- [x] Figure out if we're still supporting Internet Explorer
[No](https://github.com/HubSpot/youmightnotneedjquery/pull/264#issuecomment-1232272785)
- [x] #222

### LINKED ISSUE #272: Replace slider with radio buttons or select
We're suggesting the ES6 option be moved to a toggle separate from the IE slider (see https://github.com/HubSpot/youmightnotneedjquery/pull/271#issuecomment-1232447117)

### PR DESCRIPTION
**Pull request by Richienb**
Fixes #266, fixes #222, fixes #272, fixes #247

### CODE REVIEW

**Comment by mscuthbert** [comment]
I'm not sure that losing all the information about what works on ie8, 9, 10, etc. is a good idea, nor committing the compiled versions of the index.html / .js files -- it will mean that every commit from here on out will commit several large files to the repository.

I also don't know why the repo must be private?

Also not sure that a new build/dev system gains much -- I don't know `jade` or `gulp` but I learned them to commit to this project.  I also don't know `astro` and would have to learn it -- will installing a new development system be worth removing the accumulated knowledge of previous contributors?

-1.

**Comment by Richienb** [comment]
> I'm not sure that losing all the information about what works on ie8, 9, 10, etc. is a good idea

https://github.com/HubSpot/youmightnotneedjquery/pull/264#issuecomment-1232272785

> nor committing the compiled versions of the index.html / .js files

Yup, that unintentional.

> Also not sure that a new build/dev system gains much -- I don't know `jade` or `gulp` but I learned them to commit to this project. I also don't know `astro` and would have to learn it -- will installing a new development system be worth removing the accumulated knowledge of previous contributors?

The new build system should allow the entire website to cleanly fit within a single file. It'll be way easier to work with.

**Comment by nickserv** [comment]
I'd encourage you to read [Why Astro?](https://docs.astro.build/en/concepts/why-astro/)

> Also not sure that a new build/dev system gains much -- I don't know `jade` or `gulp` but I learned them to commit to this project. I also don't know `astro` and would have to learn it

Jade and Gulp have non-standard syntax. Jade is just a templating engine, it can't be used to easily create isolated components like Astro or React. Gulp has had pretty significant breaking changes, meanwhile Vite (which Astro uses to build JS/CSS/HTML) just had a major release with better performance and barely any breaking changes.

> will installing a new development system be worth removing the accumulated knowledge of previous contributors?

Yes because Astro uses JSX syntax (similar to React) that most JS devs are already familiar with, but without the run/build time overhead of React/Jade, while simplifying configuration and maintenance. We can also ship little to no client side JS without sacrificing user experience, which also makes browser support easier.

I just started learning Astro yesterday, and I feel I already understand it better than Gulp and Jade, which I have used for years.

**Review by nickserv — CHANGES_REQUESTED** [review]
I noticed you haven't removed Prism. Since our files are static, we should probably be able to use the `<Code />` component.

**Comment by Richienb** [comment]
> I just started learning Astro yesterday

Same, this was easier than expected

**Comment by jarikmarwede** [comment]
> > I'm not sure that losing all the information about what works on ie8, 9, 10, etc. is a good idea
> 
> [#264 (comment)](https://github.com/HubSpot/youmightnotneedjquery/pull/264#issuecomment-1232272785)

@Richienb I think you misunderstood that comment. @Abbondanzo was talking about not supporting IE in this websites code but not the code examples. I hope you understand what I mean.

**Comment by Richienb** [comment]
Oh I see. That'll need further discussion then.

My use case for usemightnotneedjquery is for discovering built-in ways to do something in the browser. I don't come here for the ponyfills and workarounds as in the older IE versions.

The examples for the older browsers have historical significance but in reality, I'll never end up using them.

**Comment by jarikmarwede** [comment]
That's why we suggested hiding the slider by default and instead adding a checkbox like "I need to support IE". The modern examples would be shown by default but the IE examples would still be there for people who need them.
see #272

**Review by Abbondanzo — CHANGES_REQUESTED** [review]
Couple o' small nits:
- The `index.html` file is generated, so we should avoid keeping that checked in. It was a source of confusion in the past as contributors would try to make changes directly to it
- Could this be broken into two PRs: one that sets up Astro and one that overhauls all the comparisons?

**Comment by Richienb** [comment]
> * The `index.html` file is generated, so we should avoid keeping that checked in. It was a source of confusion in the past as contributors would try to make changes directly to it

Yes, when I've finish the upgrade, I'm going to delete the old codebase which includes this file.

> * Could this be broken into two PRs: one that sets up Astro and one that overhauls all the comparisons?

Good idea

**Comment by nickserv** [comment]
Let me know if there's anything I can help with.

**Comment by Richienb** [comment]
Blocking, for now: https://github.com/HubSpot/youmightnotneedjquery/pull/278#issuecomment-1234957004

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -13,52 +13,35 @@ jobs:
...
-          git config --global user.name "GitHub Actions"
-          git commit -m "Deployed build artifacts (`date '+%F %T %Z'`)" || true
-          git restore .
-
-      - name: Push build artifacts to gh-pages branch
-        run: git push origin HEAD:gh-pages --force
+    steps:
+      - name: Deploy to GitHub Pages
+        id: deployment
+        uses: actions/deploy-pages@v1
```
Comment (nickserv) [inline .github/workflows/build.yml]: There's an action for Astro too https://github.com/withastro/action

**Comment by nickserv** [comment]
I think it would also be a good idea to try the [Code component](https://docs.astro.build/en/reference/api-reference/#code-), which pre renders with shiki instead of having to load prism on the client.

File: package.json [hunk 2]
```diff
@@ -1,25 +1,15 @@
...
-  "bugs": {
-    "url": "https://github.com/HubSpot/youmightnotneedjquery/issues"
+    "dev": "astro dev",
+    "build": "astro build"
   },
   "devDependencies": {
-    "gulp": "~3.3.0",
-    "gulp-jade": "~0.3.0",
-    "gulp-stylus": "0.0.7",
-    "prettier": "^2.7.1",
```
Comment (nickserv) [inline package.json:22]: You'll probably want to reinstall this and add `prettier-plugin-astro` so we can format Astro files consistently

File: package.json [hunk 3]
```diff
@@ -2,13 +2,17 @@
   "private": true,
   "scripts": {
     "dev": "astro dev",
-    "build": "astro build"
+    "build": "astro build",
+    "format": "prettier -w src astro.config.ts"
```
Comment (nickserv) [inline package.json]: I think it's better to format all files since Prettier isn't just for JavaScript
```suggestion
    "format": "prettier -w ."
```

**Comment by Richienb** [comment]
> I think it would also be a good idea to try the [Code component](https://docs.astro.build/en/reference/api-reference/#code-), which pre renders with shiki instead of having to load prism on the client.

The Prism theme that is used on the website doesn't exist in Shiki so it would require some work to port it. Prism is actually run at build-time so there is little comprimise though the code might look slightly cleaner with Shiki.

**Comment by nickserv** [comment]
Oh, in that case we can probably just import the CSS from the Prism package to make updating easier.

File: package.json [hunk 4]
```diff
@@ -1,25 +1,20 @@
...
   "scripts": {
-    "start": "gulp default",
-    "build": "gulp serve"
-  },
-  "repository": {
-    "type": "git",
-    "url": "git://github.com/HubSpot/youmightnotneedjquery.git"
-  },
-  "author": "",
-  "license": "MIT",
```
Comment (mscuthbert) [inline package.json:14]: Why is the license changing here? and the description and version disappearing?  I don't think that the license of the project can change at this point -- I wouldn't be contributing if it wasn't open source, and it seems to be a change unrelated to the scope of the PR.  Thanks!

File: src/pages/index.astro [hunk 5]
```diff
@@ -0,0 +1,542 @@
...
+              </div>
+            </section>
+          ))
+        }
+      </div>
+    </main>
+
+    <footer>
+      <p>Made by some Engineer Gamers</p>
+    </footer>
```
Comment (mscuthbert) [inline src/pages/index.astro]: Removing the credit to Adam and Zack and Hubspot also seems out of scope for the PR, unless Adam and Zack have endorsed this.

File: package.json (hunk 4 above)
Comment (Richienb) [reply package.json:14]: The license property in the package.json specifically communicates that information for npm. Since this project is not intended to be published to npm, I removed all of the metadata from the package.json file. The MIT license still exists in the LICENSE file of the repository, where you'd expect for it to be.

File: src/pages/index.astro (hunk 5 above)
Comment (Richienb) [reply src/pages/index.astro]: Yes sorry, they should still be credited. My work never had any merit for a change which is why I upgraded the message to credit a pseudonym as a joke. I still want to include that to credit community contribution but the names of the original creators shall still exist as the only proper nouns in the statement.

**Comment by Richienb** [comment]
the

**Comment by Richienb** [comment]
> I also don't know why the repo must be private?

Sorry, took me a bit to realise what you meant: by setting `"private": true`, npm knows that this is not something that is going to be published to npm and thus possibly allows us to forgo some of the metadata properties that we'd otherwise need to add which just end up being boilerplate we don't want.

**Comment by nickserv** [comment]
I have a weird intermittent build error locally. I opened #284 to enable builds on pull requests. I'm curious if the error would reproduce on GitHub Actions.
```
 error   Expected an object, got `Bud1
�ajaxmoDDblo�g1S�b�AajaxmodDblo��eb�Aajaxph1Scompeffectslg1Scomp
  �effectsmoDDblo/
  �b�AeffectsmodDblo/
elementsmoDDblo�\compelementslg1Scomp4
  �b�elementsmodDblo�\
  �b�elementsph1Scomp   �eventslg1Scomp�eventsmoDDblo��
                                                       �b�AeventsmodDblo��
                                                                          �b�Aeventsph1Scomp�utilslg1ScompGutilsmoDDbloO�T�_�AutilsmodDbloO�T�_�Autilsph1Scomp 
                                                                                                                                                               @� @ @ @
                                                                                                                                                                       E
  DSDB ` @ @ @` (string)
TypeError: Expected an object, got `Bud1
lg1Scompajaxlg1Scomp
�ajaxmoDDblo�   �b�AajaxmodDblo��eb�Aajaxph1Scompeffectslg1Scomp
�effectsmoDDblo/
�b�AeffectsmodDblo/
�b�Aeffectsph1Scompelementslg1Scomp4
elementsmoDDblo�\
�b�elementsmodDblo�\
�b�elementsph1Scomp     �eventslg1Scomp�eventsmoDDblo��
                                                       �b�AeventsmodDblo��
                                                                          �b�Aeventsph1Scomp�utilslg1ScompGutilsmoDDbloO�T�_�AutilsmodDbloO�T�_�Autilsph1Scomp 
                                                                                                                                                               @� @ @ @
                                                                                                                                                                       E
```

**Comment by mscuthbert** [comment]
> allows us to forgo some of the metadata properties that we'd otherwise need to add which just end up being boilerplate we don't want.

I guess "we don't want" is something we'll need to differ on.  If the site isn't going to be open-source licensed, I will need to stop contributing to it for various legal/contractual reasons.  That's why things like removing `"license": "MIT",` seems to me something that needs to be discussed with the community before putting into a PR that doesn't mention that as part of its remit.

**Comment by Abbondanzo** [comment]
The license is and will always be available as the [`LICENSE`](https://github.com/HubSpot/youmightnotneedjquery/blob/dev/LICENSE) file at the root of the repository. The attributes being removed from the `package.json` file here is only relevant if this project was intended to be published to NPM, which it isn't.

**Comment by Richienb** [comment]
> I have a weird intermittent build error locally. I opened #284 to enable builds on pull requests. I'm curious if the error would reproduce on GitHub Actions.

The build on my fork works:
https://github.com/Richienb/youmightnotneedjquery/runs/8175589634?check_suite_focus=true

The error could be caused by:
https://github.com/sindresorhus/map-obj/blob/4257d4dd8564efa2372c6964eb16463cb3af898d/index.js#L62

It seems that the [readFileTree call](https://github.com/Richienb/youmightnotneedjquery/blob/a5913e5f595a695198e3cf500b0a0f9c5d7d4698/src/pages/index.astro#L53) is somehow returning a string.

File: package.json [hunk 6]
```diff
@@ -1,25 +1,22 @@
...
   "devDependencies": {
-    "gulp": "~3.3.0",
-    "gulp-jade": "~0.3.0",
-    "gulp-stylus": "0.0.7",
+    "@astrojs/prism": "^1.0.1",
+    "@sindresorhus/class-names": "^2.0.0",
+    "@types/node": "^18.7.14",
+    "astro": "^1.1.3",
+    "astro-compress": "^1.0.6",
+    "map-obj": "^5.0.2",
```
Comment (Abbondanzo) [inline package.json]: I'm cool with this for now but in the future we should try to avoid adding packages that act as a thin wrapper around the methods and iterators that come out of the box. It's a level of abstraction that other developers will certainly not be familiar with and it seems like these packages are designed to provide runtime contracts for iteration. Astro files are typed so there's a compile-time guarantee that we have access to certain iterators. Awesome that you've contributed to that project before too 👍

File: src/pages/index.astro [hunk 7]
```diff
@@ -0,0 +1,722 @@
+---
+import path from 'node:path';
```
Comment (Abbondanzo) [inline src/pages/index.astro]: This is by no means a blocker to merging this PR but I'd like to keep the data processing and presentation logic separate in the future. One giant file is nice as output but more cumbersome to grok and maintain

File: src/jade/br/index.jade [hunk 8]
```diff
@@ -1,140 +0,0 @@
-doctype html
```
Comment (Abbondanzo) [inline src/jade/br/index.jade:1]: Shame to throw these out. I see that there's [an internationalization plugin for Astro](https://github.com/yassinedoghri/astro-i18next) which is nearly out of beta. I can very easily hook this site into our translation service and manually maintain a boatload of languages like Portuguese here. I'll create an issue to track this

File: src/pages/index.astro [hunk 9]
```diff
@@ -0,0 +1,722 @@
...
+                                    </a>
+                                  </li>
+                                )
+                              )}
+                            </ul>
+                          </div>
+                        )}
+                      </header>
+                      <div class="engines">
+                        {Object.entries(comparison.engines).map(
```
Comment (Abbondanzo) [inline src/pages/index.astro:189]: Since this is a centered flexbox, the item layouts look a little strange when multiple comparisons render. For example, if there are only two comparisons, it looks like this:
image
Whereas three comparisons assume a different layout:
image

File: src/pages/index.astro [hunk 10]
```diff
@@ -0,0 +1,722 @@
...
+      function resetResults() {
+        comparisonsParent.classList.remove('no_search_results');
+
+        for (const hiddenCategoryOrComparison of document.querySelectorAll(
+          '.category.hidden, .comparison.hidden'
+        )) {
+          hiddenCategoryOrComparison.classList.remove('hidden');
+        }
+
+        for (const [index, category] of document
```
Comment (Abbondanzo) [inline src/pages/index.astro:620]: Nit: this can be achieved very easily with a simple CSS selector: `:nth-child(even)` and doesn't have to be maintained here and in the search listener

**Review by Abbondanzo — COMMENTED** [review]
A handful of small comments, only a few requested changes, but overall this is looking great and nearly ready to go

File: src/pages/index.astro (hunk 9 above)
Comment (Abbondanzo) [reply src/pages/index.astro:189]: Rather than rendering all of the engines together equally, can we split out the jQuery engine to render on the left and each "comparison" to render on the right. Also would it make sense here to reintroduce the version toggle IFF the "Do you need to support IE" checkbox is true?

File: src/pages/index.astro (hunk 10 above)
Comment (Richienb) [reply src/pages/index.astro:620]: `:nth-child(even)` doesn't respect whether the child is hidden or not. This means we have to manually update this.

Comment (Abbondanzo) [reply src/pages/index.astro:620]: 🤦 I spend too much time writing React code, those elements aren't getting dropped from the DOM. Fair point!

**Review by Abbondanzo — APPROVED** [review]
:shipit: mergin'
